src/Symfony/Component/Translation/FallbackLocaleProvider.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation;
+
+use Symfony\Component\Translation\Exception\InvalidArgumentException;
+
+/**
+ * Derives fallback locales based on ICU parent locale information, by shortening locale
+ * sub tags and ultimately by going through a list of configured fallback locales.
+ *
+ * @author Matthias Pigulla <mp@webfactory.de>
+ */
+class FallbackLocaleProvider implements FallbackLocaleProviderInterface
+{
+    /**
+     * @var string[]
+     *
+     * List of fallback locales to add _after_ the ones derived from ICU information.
+     */
+    private array $ultimateFallbackLocales;
+
+    private ?array $parentLocales = null;
```

Comment:
Should this be `static`? It is used to cache parsed JSON data, so we might want some extra efficiency.

Reply:
It could be, I've no strong opinion on it. @nicolas-grekas WDYT?